### PR TITLE
Fix #4192 bug with BaseModel.construct and aliased Fields

### LINKED ISSUE #4192: BaseModel.construct does not set aliased fields consistently with standard instantiation
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.2
            pydantic compiled: False
                 install path: /Users/kyleamos/Library/Caches/pypoetry/virtualenvs/api-KwVDv1SR-py3.9/lib/python3.9/site-packages/pydantic
               python version: 3.9.1 (default, Jan 18 2022, 12:54:24)  [Clang 13.0.0 (clang-1300.0.29.30)]
                     platform: macOS-12.1-arm64-arm-64bit
     optional deps. installed: ['email-validator', 'typing-extensions']
```

```py
from pydantic import BaseConfig, BaseModel, Field


class Foo(BaseModel):
    class Config(BaseConfig):
        allow_population_by_field_name = True

    bar_: int = Field(..., alias="bar")


expected_value = 2
assert Foo(bar=expected_value).bar_ == expected_value
assert Foo.construct(bar=expected_value).bar_ == expected_value  # AttributeError: 'Foo' object has no attribute 'bar_' 
...
```

### PR DESCRIPTION
**Pull request by kylebamos**
## Change Summary

Update BaseModel.construct to consider field aliases when constructing.

## Related issue number

fix [#4192](https://github.com/samuelcolvin/pydantic/issues/4192)

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: tests/test_aliases.py [hunk 1]
```diff
@@ -335,3 +355,45 @@ def alias_generator(x):
...
+    ],
+    [
+        [False, True, 'bar', does_not_raise()],
+        [False, True, 'bar_', does_not_raise()],
+        [False, False, 'bar', does_not_raise()],
+        [False, False, 'bar_', pytest.raises(ValueError)],
+        [True, True, 'bar', does_not_raise()],
+        [True, True, 'bar_', does_not_raise()],
+        [True, False, 'bar', does_not_raise()],
+        [True, False, 'bar_', does_not_raise()],
```
Comment (kylebamos) [inline tests/test_aliases.py:375]: If construct were to behave consistently with respect to `allow_population_by_field_name_config`, this would raise a `ValueError`, which is how I had the behavior in my initial commit, but that would be a breaking change, so I decided against it.

**Comment by kylebamos** [comment]
please review

**Comment by hramezani** [comment]
Thanks @kylebamos for your patch :+1: 

Here is my understanding.
By having `Foo1` as a model with `allow_population_by_field_name=True` and `Foo2` as a model with `allow_population_by_field_name=False`:

```python
from pydantic import BaseConfig, BaseModel, Field


class Foo1(BaseModel):
    class Config(BaseConfig):
        allow_population_by_field_name = True

    bar_: int = Field(..., alias="bar")


class Foo2(BaseModel):
    class Config(BaseConfig):
        allow_population_by_field_name = False

    bar_: int = Field(..., alias="bar")


expected_value = 2
```
Here is the current state:
* Passing `alias` in `standard instantiation`:
```python
# 1 - OK
assert Foo1(bar=expected_value).bar_ == expected_value

# 2 - OK
assert Foo2(bar=expected_value).bar_ == expected_value
```
* Passing `alias` in `construct`:
```python
# 3 - AttributeError: 'Foo' object has no attribute 'bar_' -- Has to be fixed
assert Foo1.construct(bar=expected_value).bar_ == expected_value

# 4 - AttributeError: 'Foo' object has no attribute 'bar_' -- Ok
assert Foo2.construct(bar=expected_value).bar_ == expected_value
```
* Passing `field name` in `standard instantiation`:
```python
# 5 - OK
assert Foo1(bar_=expected_value).bar_ == expected_value

# 6 - ValidationError: 1 validation error for Foo2 bar -- OK
assert Foo2(bar_=expected_value).bar_ == expected_value
```
* Passing `field name` in `construct`:
```python
# 7 - OK
assert Foo1.construct(bar_=expected_value).bar_ == expected_value
# 8 - OK
assert Foo2.construct(bar_=expected_value).bar_ == expected_value
```

And your patch will fix the `Case 3` which is OK.

Regarding `Case 8`, I think it should raise an error like `Case 6`.
I think it's not related to your patch and it may break our existing implementation.
What do you think @samuelcolvin and @PrettyWood ?

File: pydantic/main.py [hunk 2]
```diff
@@ -587,10 +587,13 @@ def construct(cls: Type['Model'], _fields_set: Optional['SetStr'] = None, **valu
...
         fields_values: Dict[str, Any] = {}
         for name, field in cls.__fields__.items():
-            if name in values:
+            if field.alt_alias and field.alias in values:
+                fields_values[name] = values[field.alias]
+            elif name in values:
                 fields_values[name] = values[name]
             elif not field.required:
                 fields_values[name] = field.get_default()
+
```
Comment (hramezani) [inline pydantic/main.py:596]: Please remove this change. seems unrelated

File: tests/test_aliases.py [hunk 3]
```diff
@@ -37,7 +38,11 @@ class MyModel(BaseModel):
 
         class Config:
             alias_generator = to_upper_case
-            fields = {'my_shiny_field': 'MY_FIELD', 'foo_bar': {'alias': 'FOO'}, 'another_field': {'not_alias': 'a'}}
+            fields = {
+                'my_shiny_field': 'MY_FIELD',
+                'foo_bar': {'alias': 'FOO'},
+                'another_field': {'not_alias': 'a'},
+            }
```
Comment (hramezani) [inline tests/test_aliases.py:45]: I think these are formatting change that seems unrelated to this change. please revert them and all the other formatting changes.

File: tests/test_aliases.py [hunk 4]
```diff
@@ -335,3 +355,45 @@ def alias_generator(x):
...
+def test_allow_population_by_field_name_config(
+    use_construct: bool,
+    allow_population_by_field_name_config: bool,
+    arg_name: str,
+    expectation: ContextManager,
+):
+    expected_value: int = 7
+
+    class Foo(BaseModel):
+
```
Comment (hramezani) [inline tests/test_aliases.py:387]: Please remove this blank line

**Comment by hramezani** [comment]
Please update

File: pydantic/main.py (hunk 2 above)
Comment (kylebamos) [reply pydantic/main.py:596]: Will do, black did this when I ran it and I don't think it picked up on the local config file for whatever reason, but I didn't make a dedicated python env just for this when I cloned it so it's probably my fault.

**Comment by kylebamos** [comment]
please review

**Comment by samuelcolvin** [comment]
@kylebamos I have to say I find your avatar very confusing, I keep thinking there are lots of negative comments and reviews.

I'm marginally pro this change, my concern is whether it could be a breaking change to anyone. @PrettyWood do you have any thoughts?

Also conflicts needs solving, please update.

**Comment by kylebamos** [comment]
Conflict resolved. Avatar fixed to your liking, I hope.

**Comment by kylebamos** [comment]
Please review

**Comment by samuelcolvin** [comment]
> Avatar fixed to your liking, I hope.

Much better :+1: I'm sure I'm not the only one who will prefer it :smirk:.

**Comment by PrettyWood** [comment]
LGTM! Good for v1.10!

**Comment by samuelcolvin** [comment]
@kylebamos thanks so much for this.
